api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java

```diff
@@ -408,6 +408,54 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
       return ROWS_MIGHT_MATCH;
     }
 
+    @Override
+    public <T> Boolean notStartsWith(BoundReference<T> ref, Literal<T> lit) {
+      Integer id = ref.fieldId();
+
+      // Iceberg does not implement SQL 3-boolean logic. Return ROWS_MIGHT_MATCH
+      // to allow the query engine to make its own decision regarding NULL values.
+      if (mayContainNull(id)) {
+        return ROWS_MIGHT_MATCH;
```

Comment:
I don't think this is correct. In SQL, `null LIKE 'a%'` should result in `null`, and `null NOT LIKE 'a%'` should also result in `null`. So in both cases, `null` does not match and is ignored.

Here, I think this should use `containsNullsOnly(id)`. In that case, we know that no rows match and can exit before checking bounds.

Reply:
Sure. Originally, when this was worked on, you had said to leave it up to the engine to decide because we don't implement SQL 3-boolean logic.

I'm ok with your argument presented here though. Have any input on how to update the comment then?

Reply:
I'm not sure what I was thinking at the time, but I'm pretty sure it was wrong! No need to mention SQL boolean logic. These classes implement 2-valued logic correctly and we handle SQL translation.

Reply:
Actually, this was my mistake.

Iceberg expressions implement normal logic and ignore 3-value logic. null does not start with any prefix. If there is a null value then it does not start with the given prefix and would match the expression. That means that `mayContainNull` should cause this to return `ROWS_MIGHT_MATCH` as you have here.

If SQL chooses to filter out those rows with 3-value logic, then that's okay. But Iceberg should return them because they match the Iceberg expression. I think what I was confused about was the mention of the 3-value logic here. I would probably remove that because the semantics of Iceberg expressions should not change, as is implied by the statement, "return ROWS_MIGHT_MATCH to allow the query engine to make its own decision". Instead, I'd change the comment to something like "null does not start with any prefix, any null row matches".

Reply:
How does this look?

```java
      // Iceberg expressions implement normal logic and ignore 3-value logic. Null does not start with any prefix.
      // If there is a null value then, that value doesn't start with the given prefix and would match the expression.
      if (mayContainNull(id)) {
        return ROWS_MIGHT_MATCH;
      }
```

Reply:
I'd probably just remove the comment. It makes sense that a null value doesn't match a prefix.

Reply:
Removed the comment entirely.